### PR TITLE
[SPARK-36431][SQL] Support TypeCoercion of ANSI intervals with different fields

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
 Support TypeCoercion of ANSI intervals with different fields

### Why are the changes needed?
 Support TypeCoercion of ANSI intervals with different fields

### Does this PR introduce _any_ user-facing change?
After this pr user can 
 - use comparison function with  different fields of DayTimeIntervalType/YearMonthIntervalType such as `INTERVAL '1' YEAR` > `INTERVAL '11' MONTH`
 - support different field of ansi interval type in collection function such as `array(INTERVAL '1' YEAR, INTERVAL '11' MONTH)`
 - support different field of ansi interval type in `coalesce` etc..

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @MaxGekk

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by MaxGekk** [comment]
@AngersZhuuuu Thank you for the ping. Unfortunately, I cannot review it at the moment. @cloud-fan @sarutak @beliefer @Peng-Lei Could you review this PR, please.

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql [hunk 1]
```diff
@@ -322,3 +322,8 @@ SELECT INTERVAL '153722867280' MINUTE;
 SELECT INTERVAL '-153722867280' MINUTE;
 SELECT INTERVAL '54.775807' SECOND;
 SELECT INTERVAL '-54.775807' SECOND;
+
+SELECT INTERVAL '1' DAY > INTERVAL '1' HOUR;
+SELECT INTERVAL '1' YEAR = INTERVAL '1' MONTH;
+SELECT INTERVAL '1-1' YEAR TO MONTH = INTERVAL '13' MONTH;
+SELECT INTERVAL '13' MONTH > INTERVAL '1' YEAR;
```
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/inputs/interval.sql:329]: Could you add more tests:
- At least, every supported ANSI interval type should be tested
- More compare ops
- Negative, 0, positive intervals
- Negative tests
- Comparison of year-month with day-time intervals (should fail).

Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/inputs/interval.sql]: Update a lot test, should be enough

**Comment by sarutak** [comment]
@AngersZhuuuu The GA failure was resolved in #33660. Could you rebase to `master`?

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu The GA failure was resolved in #33660. Could you rebase to `master`?

Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 2]
```diff
@@ -840,10 +840,17 @@ abstract class BinaryComparison extends BinaryOperator with Predicate {
 
   final override val nodePatterns: Seq[TreePattern] = Seq(BINARY_COMPARISON)
 
-  override def checkInputDataTypes(): TypeCheckResult = super.checkInputDataTypes() match {
-    case TypeCheckResult.TypeCheckSuccess =>
-      TypeUtils.checkForOrderingExpr(left.dataType, this.getClass.getSimpleName)
-    case failure => failure
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val matched = (left.dataType, right.dataType) match {
+      case (l: DayTimeIntervalType, r: DayTimeIntervalType) => TypeCheckResult.TypeCheckSuccess
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala:845]: It's a bit weird that we allow different types in binary comparison. Can we fix the type coercion instead? e.g. `TypeCoercion.findTightestCommonType`. This is also more general, we can also support `coalesce(interval1, interval2)`

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: > It's a bit weird that we allow different types in binary comparison. Can we fix the type coercion instead? e.g. `TypeCoercion.findTightestCommonType`. This is also more general, we can also support `coalesce(interval1, interval2)`

Hmm like current?

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: > It's a bit weird that we allow different types in binary comparison. Can we fix the type coercion instead? e.g. `TypeCoercion.findTightestCommonType`. This is also more general, we can also support `coalesce(interval1, interval2)`

If not change here, `checkInputDataTypes` will be false and expression is not resolved for  below UT.
```
 checkEvaluation(EqualTo(
      Literal.create(10,
        YearMonthIntervalType(YearMonthIntervalType.YEAR, YearMonthIntervalType.MONTH)),
      Literal.create(10,
        YearMonthIntervalType(YearMonthIntervalType.MONTH, YearMonthIntervalType.MONTH))), true)
```
If reasonable? or should I just remove the UT

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala [hunk 3]
```diff
@@ -120,6 +120,9 @@ object AnsiTypeCoercion extends TypeCoercionBase {
     case (_: TimestampType, _: DateType) | (_: DateType, _: TimestampType) =>
       Some(TimestampType)
 
+    case (_: DayTimeIntervalType, _: DayTimeIntervalType) => Some(DayTimeIntervalType())
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala:123]: shall we pick the smaller start field and larger end field as the result fields?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 4]
```diff
@@ -867,6 +869,9 @@ object TypeCoercion extends TypeCoercionBase {
       case (_: TimestampType, _: DateType) | (_: DateType, _: TimestampType) =>
         Some(TimestampType)
 
+      case (_: DayTimeIntervalType, _: DayTimeIntervalType) => Some(DayTimeIntervalType())
+      case (_: YearMonthIntervalType, _: YearMonthIntervalType) => Some(YearMonthIntervalType())
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:873]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 5]
```diff
@@ -656,6 +656,8 @@ abstract class TypeCoercionBase {
       (leftType, rightType) match {
         case (_: DecimalType, NullType) => true
         case (NullType, _: DecimalType) => true
+        case (_: DayTimeIntervalType, _: DayTimeIntervalType) => true
+        case (_: YearMonthIntervalType, _: YearMonthIntervalType) => true
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:660]: why do we need this? the default case should work for interval types

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/IntervalExpressionsSuite.scala [hunk 6]
```diff
@@ -661,4 +661,17 @@ class IntervalExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
         }
     }
   }
+
+  test("SPARK-36431: Support comparison of ANSI intervals with different fields") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/IntervalExpressionsSuite.scala:665]: The UT here should not test type coercion. Let's remove it.

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql [hunk 7]
```diff
@@ -322,3 +322,28 @@ SELECT INTERVAL '153722867280' MINUTE;
 SELECT INTERVAL '-153722867280' MINUTE;
 SELECT INTERVAL '54.775807' SECOND;
 SELECT INTERVAL '-54.775807' SECOND;
+
+SELECT INTERVAL '1' DAY > INTERVAL '1' HOUR;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/interval.sql:326]: let's focus on different fields, not different binary operators.

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/interval.sql:326]: and test `array(...)`, `coalesce(...)` as well.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 5 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Removed

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/IntervalExpressionsSuite.scala (hunk 6 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/IntervalExpressionsSuite.scala]: Done

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql (hunk 7 above)
Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/inputs/interval.sql:326]: Done

Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/inputs/interval.sql:326]: > let's focus on different fields, not different binary operators.

I think current can cover all the case?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 4 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala [hunk 8]
```diff
@@ -120,6 +121,15 @@ object AnsiTypeCoercion extends TypeCoercionBase {
     case (_: TimestampType, _: DateType) | (_: DateType, _: TimestampType) =>
       Some(TimestampType)
 
+    case (t1: DayTimeIntervalType, t2: DayTimeIntervalType) =>
+      Some(DayTimeIntervalType(
+        IntervalUtils.minAnsiIntervalField(t1.startField, t2.startField),
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala:126]: just `t1.startField.min(t2.startField)`?

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala]: > just `t1.startField.min(t2.startField)`?

There is such a convenient API。。。Updated

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql [hunk 9]
```diff
@@ -322,3 +322,47 @@ SELECT INTERVAL '153722867280' MINUTE;
...
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/interval.sql:326]: I'd only keep
```
SELECT INTERVAL '1' DAY > INTERVAL '1' HOUR;
SELECT INTERVAL '1 02' DAY TO HOUR = INTERVAL '02:10:55' HOUR TO SECOND;
SELECT INTERVAL '1' YEAR < INTERVAL '1' MONTH;
SELECT INTERVAL '-1-1' YEAR TO MONTH = INTERVAL '-13' MONTH;

SELECT array(INTERVAL '1' YEAR, INTERVAL '1' MONTH);
SELECT array(INTERVAL '1' DAY, '01:01' HOUR TO MINUTE);
SELECT coalesce(INTERVAL '1' YEAR, INTERVAL '1' MONTH);
SELECT coalesce(INTERVAL '1' DAY, '01:01' HOUR TO MINUTE); 
```

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/interval.sql:326]: You can test more combinations in `TypeCoercionSuite`

Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/inputs/interval.sql:326]: > You can test more combinations in `TypeCoercionSuite`

Done

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql [hunk 10]
```diff
@@ -322,3 +322,13 @@ SELECT INTERVAL '153722867280' MINUTE;
...
+SELECT INTERVAL '1 02' DAY TO HOUR = INTERVAL '02:10:55' HOUR TO SECOND;
+SELECT INTERVAL '1' YEAR < INTERVAL '1' MONTH;
+SELECT INTERVAL '-1-1' YEAR TO MONTH = INTERVAL '-13' MONTH;
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/interval.sql:329]: let's add a negative test as well
```
SELECT INTERVAL 1 MONTH > INTERVAL 20 DAYS;
```

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql [hunk 11]
```diff
@@ -322,3 +322,13 @@ SELECT INTERVAL '153722867280' MINUTE;
...
+
+SELECT array(INTERVAL '1' YEAR, INTERVAL '1' MONTH);
+SELECT array(INTERVAL '1' DAY, '01:01' HOUR TO MINUTE);
+SELECT coalesce(INTERVAL '1' YEAR, INTERVAL '1' MONTH);
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/interval.sql:333]: ditto

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql (hunk 10 above)
Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/inputs/interval.sql:329]: Done

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql (hunk 11 above)
Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/inputs/interval.sql:334]: Done

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql [hunk 12]
```diff
@@ -322,3 +322,15 @@ SELECT INTERVAL '153722867280' MINUTE;
...
+SELECT INTERVAL '1 02' DAY TO HOUR = INTERVAL '02:10:55' HOUR TO SECOND;
+SELECT INTERVAL '1' YEAR < INTERVAL '1' MONTH;
+SELECT INTERVAL '-1-1' YEAR TO MONTH = INTERVAL '-13' MONTH;
+SELECT INTERVAL 1 MONTH > INTERVAL 20 DAYS;
+
+SELECT array(INTERVAL '1' YEAR, INTERVAL '1' MONTH);
+SELECT array(INTERVAL '1' DAY, INTERVAL '01:01' HOUR TO MINUTE);
+SELECT coalesce(INTERVAL '1' YEAR, INTERVAL '1' MONTH);
+SELECT coalesce(INTERVAL '1' DAY, INTERVAL '01:01' HOUR TO MINUTE);
+SELECT coalesce(INTERVAL 1 MONTH, INTERVAL 20 DAYS);
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/interval.sql:336]: don't miss the `array(...)` test case

Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/inputs/interval.sql:336]: Done

**Review by cloud-fan — APPROVED** [review]

**Review by MaxGekk — APPROVED** [review]
+1, LGTM. Merging to master/3.2.
Thank you, @AngersZhuuuu and @cloud-fan for review.

**Pull request closed without merging**